spec/formatting.md

```diff
@@ -222,6 +222,74 @@ the following steps are taken:
 
    The form that resolved _operand_ and _option_ values take is implementation-defined.
 
+   A _local-declaration_ binds the resolved value of an _expression_
+   to a _variable_.
+   Thus, the output of one _function_ is potentially the _operand_
+   of another _function_,
+   or the value of one of the _options_ for another function.
+   For example, in
+   ```
+   .input {$n :number minIntegerDigits=3}
+   .local $n1 = {$n :number maxFractionDigits=3}
+   ```
+   the output of the first call to `:number`
+   is the input of the second call to `:number`.
+
+   Implementations that provide a means for defining custom functions
+   SHOULD provide a means for function implementations
+   to return values that contain enough information
+   (e.g. the resolved _operand_ and _option_ values
+   that the function was called with)
+   to be used as arguments to subsequent calls
+   to the function implementations.
+   For example, a MessageFormat implementation in a typed programming language
+   MAY define an interface for custom function implementations.
+   Such an interface SHOULD define an implementation-specific
+   argument type `T` and return type `U`
+   for implementations of formatting functions
+   such that `U` can be coerced to `T` without loss of information.
+   The type `U`
+   (or a type that `U` can be coerced to without loss of information)
+   SHOULD also be the input type of implementations of
+   custom selector functions.
```

Comment:
**Blocking:**
The explicit references to "formatting functions" and "selector functions" really do need to get dropped. That also allows the second sentence here to be dropped.

Consider this message:
```
.input {$n :number}
.match {$n}
one {{{$n} is one!!1}}
* {{{$n} is not one.}}
```

Here, the resolved value of `{$n :number}` is used both for selection and formatting. So is the `:number` implementation a "formatting function", or a "selector function"? It is neither, it's just a "function", or if we really must add a qualifier, it's a "message function".

To be clear, it is possible for an implementation to implement this using two separate "formatter" and "selector" functions, but the spec does not and should not mandate this. Instead, such an implementation carries an extra burden e.g. ensuring that any Resolution Error in the resolution of `{$n :number}` is only emitted once, rather than separately for the selector and formatter.

```suggestion
   Such an interface SHOULD define an implementation-specific
   argument type `T` and return type `U`
   for implementations of functions
   such that `U` can be coerced to `T` without loss of information.
```

Reply:
I'm not super comfortable with this change because it implies that a selector has to have a return type that's the same as whatever a formatter returns. At the implementation level, a selector returns a list of keys. I'm not sure how the same implementation (as opposed to two different implementations that happen to be registered to the same name in the function registry) can return both. I haven't looked at other implementations closely besides my own ICU4C one and ICU4J, but in those two implementations, it's clear that these are really different interfaces.

However, I guess one reading of your suggested text is that functions that don't return any value can be defined, so I guess that's compatible at least with my own implementation, where selectors are passed a list of keys to fill in as an out-parameter, and return `void`.